R/metrics-sample.R

```diff
@@ -230,16 +245,87 @@ dss_sample <- function(observed, predicted, ...) {
 #' Forecasts with scoringRules, <https://www.jstatsoft.org/article/view/v090i12>
 #' @keywords metric
 
-crps_sample <- function(observed, predicted, ...) {
+crps_sample <- function(observed, predicted, separate_results = FALSE, ...) {
   assert_input_sample(observed, predicted)
 
-  scoringRules::crps_sample(
+  crps <- scoringRules::crps_sample(
     y = observed,
     dat = predicted,
     ...
   )
+
+  if (separate_results) {
+    medians <- apply(predicted, 1, median)
+    dispersion <- scoringRules::crps_sample(
+      y = medians,
+      dat = predicted,
+      ...
+    )
+    overprediction <- rep(0, length(observed))
+    underprediction <- rep(0, length(observed))
+
+    overprediction[observed < medians] <- scoringRules::crps_sample(
+      y = observed[observed < medians],
+      dat = predicted[observed < medians, , drop = FALSE],
+      ...
+    )
+    underprediction[observed > medians] <- scoringRules::crps_sample(
+      y = observed[observed > medians],
+      dat = predicted[observed > medians, , drop = FALSE],
+      ...
+    )
+
+    overprediction[overprediction > 0] <-
+      overprediction[overprediction > 0] - dispersion[overprediction > 0]
```

Comment:
same is true here a lot of replication and over/under are independent of each other so a lot of wasted compute here.

Reply:
I agree (same for WIS which is calculate three times for getting overpred, underpred and dispersion). It would be good to think of a set up for both of these together in a separate issue, I think.

Reply:
Added an issue: https://github.com/epiforecasts/scoringutils/issues/870